ha/core/controllers/pcs/node_controller.py

```diff
@@ -152,6 +149,119 @@ def status(self, nodeids: list = None) -> dict:
         """
         raise HAUnimplemented("This operation is not implemented.")
 
+    def _validate_permissions(self) -> None:
+
+        # confirm that user is root or part of haclient group"
+        user = getpass.getuser()
+        group_id = os.getgid()
+
+        try:
+            # find group id for root and haclient
+            id_ha = grp.getgrnam(const.USER_GROUP_HACLIENT)
+            id_root = grp.getgrnam(const.USER_GROUP_ROOT)
+
+            # if group not root or haclient return error
+            if group_id != id_ha.gr_gid and group_id != id_root.gr_gid:
+                Log.error(f"User {user} does not have necessary permissions to execute this CLI")
+                raise HAInvalidPermission(
+                            f"User {user} does not have necessary permissions to execute this CLI")
+
+            # The user name "hauser"; which is part of the "haclient" group;
+            # is used by HA.
+            # internal commands are allowed only if the user is "hauser"
+            # As of now, every HA CLI will be internal command. So, we
+            # do not need this change. We can revisit this if needed in future
+            #if user == const.USER_HA_INTERNAL:
+            #    self._is_hauser = True
+
+
+        # TBD : If required raise seperate exception  for root and haclient
+        except KeyError:
+            Log.error("Group root / haclient is not defined")
+            raise HAInvalidPermission("Group root / haclient is not defined ")
+
+    def check_cluster_feasibility(self, nodeid: str) -> dict:
+        """
+            Check whether the cluster is going to be offline after node with nodeid is stopped.
+        Args:
+            nodeid (str): Private fqdn define in conf store.
+        Returns:
+            Dictionary : {"status": "", "msg":""}
+        """
+        self.validate_node(node_id=nodeid)
+        offline_nodes, node_list = self._get_offline_nodes()
+        if (len(offline_nodes) + 1) >= ((len(node_list)//2) + 1):
+            return {"status": const.STATUSES.FAILED.value, "error": "Stopping the node will cause a loss of the quorum"}
+        else:
+            return {"status": const.STATUSES.SUCCEEDED.value, "error": ""}
+
+    def validate_node(self, node_id: str) -> bool:
+        """
+        Check node is a part of cluster
+        Args:
+            node_id ([str]): Private fqdn define in conf store.
+        Raises: ClusterManagerError
+        Returns: bool
+        """
+        _, node_list = self._get_offline_nodes()
+        if node_id in node_list:
+            return True
+        raise ClusterManagerError(f"nodeid = {node_id} is not a valid node id")
+
+    def _get_offline_nodes(self):
+        """
+        Get list of offline nodes ids.
```

Comment:
Add the return values in description. In all functions describe the passed arguments and expected return values

Reply:
Added details in all places where required.

Reply:
Do not see the details in latest patch